src/compiler/checker.ts

```diff
@@ -8844,35 +9127,58 @@ module ts {
 
         /** Check a decorator */
         function checkDecorator(node: Decorator): void {
-            let expression: Expression = node.expression;
-            let exprType = checkExpression(expression);
-
+            let signature = getResolvedSignature(node);
+            let returnType = getReturnTypeOfSignature(signature);
+            if (returnType.flags & TypeFlags.Any) {
+                return;
+            }
+            
+            let expectedReturnType: Type; 
+            let diagnosticChainHead: DiagnosticMessageChain;
+            let decoratorKind: string;
             switch (node.parent.kind) {
                 case SyntaxKind.ClassDeclaration:
                     let classSymbol = getSymbolOfNode(node.parent);
                     let classConstructorType = getTypeOfSymbol(classSymbol);
-                    let classDecoratorType = instantiateSingleCallFunctionType(getGlobalClassDecoratorType(), [classConstructorType]);
-                    checkTypeAssignableTo(exprType, classDecoratorType, node);
+                    expectedReturnType = getUnionType([classConstructorType, voidType]);
+                    decoratorKind = "class";
+                    break;
+
+                case SyntaxKind.Parameter:
+                    expectedReturnType = voidType;
+                    decoratorKind = "parameter";

```

Comment:
Also, technically inserting these names is not localization friendly because how the rest of the message is translated depends on these words. For example, consider what would happen in English if one of these words began with a vowel, like `interface` or `enum`:
> The return type of **a 'enum' decorator**...

Should be "an enum decorator"